### PR TITLE
Restore the MSVC port of OCaml

### PR DESCRIPTION
**Pull request by dra27**
_None of this is yet ready for actual code review - in particular, quite a few commits intentionally make OCaml MSVC-only!_ It does, however, work, and I therefore think it's worth sharing, even if only to put it in the 5.1 milestone and to expose the task list publicly. Any and all questions about this very much welcomed, of course!

Last week, Microsoft added in [C11 atomics in Visual Studio 2022](https://devblogs.microsoft.com/cppblog/c11-atomics-in-visual-studio-2022-version-17-5-preview-2/) and this PR restores the MSVC port of OCaml on this basis. At the moment, 32-bit bytecode does not work, but I think this may be down a miscompilation in cl. 64-bit, however, is completely working and the testsuite on my machine in both bytecode and native code is fully passing.

I have disabled CI, with the intention of gradually re-enabling it as parts of this get carved out in separate PRs and the MSVC-only hacks get put into their proper places. In particular, rebasing and manually testing this PR should help to prevent further divergence affecting this port, even if it can't yet be fully re-enabled.

I've done a side-by-side disassembly of `amd64.o` and `amd64nt.obj`, so while the actual assembler file needs considerably tidying, I believe it is correct and that the remaining bug affecting native code is actually in ocamlopt, rather than the runtime.

- [x] `Iextcall` to non-alloc functions fixed (fairly sure this is affecting mingw-w64 in 5.0, but not as readily)
- [ ] MSVC atomics working as expected, rather than "as working"!
- [ ] 32-bit bytecode support (almost certainly MSVC bug(s)?)
- [ ] Macros for `__declspec(thread)` (possibly via C11 `threads.h`?)
- [ ] winpthreads integration by submodule (or possibly via C11 `threads.h`?)
- [ ] Eliminate C compiler warnings (fairly sure these are atomics bugs in MSVC)
- [ ] Formalise or eliminate the addition of `-nodefaultlib:LIBCMT`
- [ ] Fix the additional inclusions of `winbase.h` (affects mingw-w64 as well)
- [ ] Fix reliance on `sys/time.h` (affects mingw-w64 as well)
- [ ] Debugging parts (`-Od`, etc.) removed
- [ ] amd64nt.asm either eliminated (by using ocamlopt's DSL for the amd64 runtime support object) or updated to be slightly less-unmaintainable (either using MASM macros properly or switching the amd64.S to Intel syntax and generating all from a common base)
- [ ] Name-mangling properly fixed for MASM-compatibility
- [ ] Update `README.win32.adoc`

### CODE REVIEW

File: runtime/platform.c [hunk 1]
```diff
@@ -209,23 +211,29 @@ void caml_mem_decommit(void* mem, uintnat size)
...
-#ifdef DEBUG
-  caml_lf_skiplist_remove(&mmap_blocks, (uintnat)mem);
-#endif
 }
 
 #define Min_sleep_ns       10000 // 10 us
 #define Slow_sleep_ns    1000000 //  1 ms
 #define Max_sleep_ns  1000000000 //  1 s
 
+/* Again, TODO: how does this work on mingw-w64 atm? */
```
Comment (nojb) [inline runtime/platform.c:223]: Mingw-w64 emulation is at:

https://github.com/mirror/mingw-w64/blob/master/mingw-w64-crt/misc/mingw_usleep.c

If I understand correctly, we would include some equivalent code to have it available for MSVC as well?

Comment (dra27) [reply runtime/platform.c:223]: Thanks for digging! TIL there are many more emulated CRT functions in mingw-w64! Indeed, this looks like something we'd want to move to win32.c as `caml_usleep` and then have `osdeps.h` (I think?) deal with it.

In the absence of benchmarks, I'm not sure what's worse - `usleep` in mingw-w64 is actually "millisecond sleep", whereas I'm not sure that the function I put in platform.c (which I very quickly grabbed from somewhere!) making that many system calls is going to be particularly good at sub-ms granularity). On the other hand, this is used for `SPIN_WAIT` throughout the gc.

Comment (XVilka) [reply runtime/platform.c:223]: The most precise way to sleep on Windows is this: https://stackoverflow.com/questions/7827062/is-there-a-windows-equivalent-of-nanosleep

Comment (dra27) [reply runtime/platform.c:223]: Thanks @XVilka. though the issue is not the _precision_ but the _granularity_.

**Comment by dra27** [comment]
This PR has run its course - the reviewable work continues in #12954!

**Pull request closed without merging**